Performance optimization - Iterator

Ichoran opened the pull request:
Rewrite of `span` to avoid `.buffered` and to only use `mutable.Queue` when absolutely necessary.  Avoid `buffered` in `dropWhile` and `collect`.

Tidied up other methods by using local `return`.  This doesn't always improve performance, but the code is more compact and (I'd argue) easier to follow, and the bytecode is reduced which in appropriate contexts will help with inlining.

Summary of improvements (all timings JDK 1.8.0_25 Linux):

```
method        reason
===========   ===============================================================
commit 1
-----------   ---------------------------------------------------------------
flatMap       34 bytes instead of 62 in hasNext 
find          41 bytes instead of 53
indexWhere    1.5x faster on small collections (some contexts)
indexOf       89 bytes instead of 110 

commit 2
-----------   ---------------------------------------------------------------
collect       2.3x faster on small collections, 1.5x on large

span          1.6-1.7x faster on small collections
              0.85x-1.8x slower/faster on large collections
                depending on how much must be cached (0.85x all, 1.8x none)

dropWhile     1.2x faster on small collections, half the garbage
```

som-snytt commented:
Don't fear the local return!

or maybe

The return of the local return!

I can't decide which t-shirt to order.

mpociecha commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -405,6 +413,7 @@ trait Iterator[+A] extends TraversableOnce[A] {
    *  @note    Reuse: $consumesAndProducesIterator
    */
   def filter(p: A => Boolean): Iterator[A] = new AbstractIterator[A] {
+    // TODO 2.12 - Make a full-fledged FilterImpl that will reverse sense of p
```
Maybe wouldn't it be bad idea to have JIRA issues created for such TODOs? Then you could add here an issue number. WDYT?

Ichoran replied:
I'd agree if I wasn't going to go fix these in 2.12.x right after 2.11.x is out.  As it is, it's faster and easier to grep for "TODO 2.12", and it's kind of an awkward fit for JIRA since the original improvements weren't listed there.

adriaanm commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -470,13 +479,22 @@ trait Iterator[+A] extends TraversableOnce[A] {
+        }
+        else hdStatus = -1
+      }
+      hdStatus > 0
```
Did you mean `hdStatus >= 0`? If not, `hdStatus == 1` is clearer.

While you're at it, I'd suggest

```
final val NO_NEXT   = -1
final val HAS_NEXT  = 0
final val FOUND     = 1
```

Ichoran replied:
I did mean `> 0` because `hasNext`'s job is to find the next thing, and if it's not found, there isn't a `next`.  I think I had a fourth status code originally, which is why it was `> 0`; I'll fix it.  I will check to make sure the compiler and/or JVM manages to turn `final val` into inlined constants.

adriaanm replied:
Ok! The type checker will constant fold final vals without return types.

adriaanm commented:
Could you re-group the changes into commits consisting solely of
-  "obvious" rewrite 
-  significant rewrite (the ones using status codes)

... and assess the performance impact individually? If the bigger rewrites don't come with similarly sized perf boosts, I'm inclined to be conservative about them.

Given that this needs some more work, I'm open to delaying (the more significant rewrite) to 2.11.7.

Ichoran commented:
Sure, I'll run the benchmarks again and put them up here or on the ticket, but the big rewrites were done precisely because it makes a relatively big difference in at least some use cases.  I probably should re-rewrite `collect` to cache the result of `applyOrElse`, also.

Ichoran committed:
Performance optimization - Iterator flatMap, find, indexWhere, indexOf

Tightened up bytecode, logic, and/or performance by using local return instead of a mutable variable in several methods.

Performance/bytecode size improvements (smaller bytecode = better inlining)

```
method        reason
===========   ===============================================================
flatMap       hasNext bytecode 34 bytes down from 62
find          bytecode 41 bytes instead of 53
indexWhere    1.5x faster on small collections (some contexts)
indexOf       bytecode 89 bytes instead of 110
```

Ichoran committed:
Performance optimization - Iterator span, collect, dropWhile

Rewrite of span to avoid double-indirection of `.buffered` and to avoid use of `mutable.Queue` unless it is absolutely necessary.  Rewrite of `span` and `dropWhile` to also avoid `.buffered` (less DRY but single vs. double indirection and object allocation).

Performance improvements:

```
method        reason
===========   ===============================================================
collect       2.3x faster on small collections, 1.5x on large

span          1.6-1.7x faster on small collections
              0.85x-1.8x slower/faster on large collections
                depending on how much must be cached (0.85x all, 1.8x none)

dropWhile     1.2x faster on small collections, half the garbage
```

Ichoran commented:
@adriaanm - Recommitted in the way you said and did my best to remove magic constants, but Scala really doesn't like low-overhead magic constants so I couldn't without changes that are probably not binary compatible.  I can't find any way to get a static val without a companion object, but the anonymous inner classes don't have companion objects now.  def will be inlined with an annotation but only when -optimize is passed, which seems a little dangerous (and also clutters up the classfile with method names for the constants).  I didn't benchmark how defs fare without optimization; maybe I should.  Better would be for compiler support for less painful ways to get constants in anonymous classes.

adriaanm commented:
final vals without return type will be constant folded by the type checker, so guaranteed to have the constant itself in bytecode:

`$ cat /tmp/bla.scala`

```
class C {
  private final val myConst = 2
  def foo = println(myConst)
}
```

`$ scalac -Xprint:typer /tmp/bla.scala`

```
[[syntax trees at end of                     typer]] // bla.scala
package <empty> {
  class C extends scala.AnyRef {
    def <init>(): C = {
      C.super.<init>();
      ()
    };
    final private[this] val myConst: Int(2) = 2;
    final <stable> <accessor> private def myConst: Int(2) = 2;
    def foo: Unit = scala.this.Predef.println(2)
  }
}
```

adriaanm commented:
Moving to 2.11.8 since it needs a rebase.

SethTisue commented:
@Ichoran ping

retronym commented:
> final vals without return type will be constant folded by the type checker, so guaranteed to have the constant itself in bytecode:

Unfortunately these can't be defined locally to a method. While we could move the ones defined in this PR to a private object, I'm happy to let this one go through as is.

LGTM

SethTisue commented:
@Ichoran I tried to merge this onto 2.12.x, but there were merge conflicts with a couple of @som-snytt's commits (5ebb7bb3bfc0e5152a707e693379d5ac5b5d9355 and 5b02bb4f8993f8e09ff1223fbdd4acf78eabd964). you can make an additional PR that does the merge? (you'll want to either base it off #4732, or wait for #4732 to be merged)

som-snytt commented:
I wonder if restoring the comment "Giving a name..." L605 was unintended? For some reason, the words confuse me. And there are so many of them.

It would be neat to have a macro to do the lite-weight enumeration.

```
var status = n"Seek,Found,Empty" ; status = n"Found" ; status = n"Fund" // oops
```

Ichoran commented:
@SethTisue - I think the new span supercedes the other patches, but I'll merge it when 4732 is in.  If I make a new branch and pull in this patch and fix the merge conflicts, then submit a PR against 2.12.x, that will resolve the issue fully, I hope?  I'm not completely sure how the 2.11.x -> 2.12.x forwarding process works (e.g. how automatic it is), and I'd like to ensure that this isn't an ongoing maintenance headache.

som-snytt commented:
@Ichoran Sorry about that. Usually I avoid messing with the timeline. From the ticket SI-9332, I see it was Adriaan invited a PR. How could I resist? He suggested I was compelled by logical necessity.

It looks like the first patch was correctness, and the second was actually performance, not to push everything through the queue. It would have been fun to see the microbenchmark against that version.

SethTisue commented:
#4732 is now merged.

@Ichoran yes, the process is as you describe (assuming by "pull in" you mean `git merge` and not `git cherry-pick`). we don't have any special automation for 2.11.x->2.12.x merges, we just do an ordinary `git merge` manually from time to time.

adriaanm commented:
@som-snytt I knew I could charm you with my soundness. No regrets -- YO[C](https://www.google.com/search?q=define%20code)O

adriaanm commented:
regarding pulling in, `git pull scala 2.12.x` will do the merge into your current branch (`--rebase` makes it a rebase), assuming a prior `hub remote add scala`

SethTisue commented:
@Ichoran nbd, but just a ping on this since M3 is coming up in ~2 weeks

Ichoran commented:
@SethTisue - I am not quite sure what the status of things are supposed to be, but I am manually merging in som's changes, which looks like what is necessary on the 2.12.x branch at this point.  All of my changes seem to be there.

Ichoran commented:
@SethTisue - I'm completely confused.  Aside from som's comment, which is much better and briefer than mine, the merge seems to already have happened.  I've tried to figure out what you wanted me to do or how there could have been a problem, but i don't see it.  blame indicates that som's patch was applied first, then mine.  Is there still anything for me to do?  If so, I can't find it, so can you point out really clearly in the 2.12 branch exactly what's missing?  Sorry if I'm being dense.

SethTisue commented:
@Ichoran looks like @lrytz took care of it in https://github.com/scala/scala/pull/4761, 4 days ago? if what he did to resolve the merge conflict looks good to you, then we're done.

Ichoran commented:
@SethTisue - Yes, I've checked it all preposterously carefully now, since I was assuming I had something to do.
